clippy_lints/src/redundant_clone.rs

```diff
@@ -0,0 +1,287 @@
+                || (match_def_path(cx.tcx, fn_def_id, &paths::TO_STRING_METHOD)
+                    && match_type(cx, arg_ty, &paths::STRING));
+
+            let from_deref = !from_borrow
+                && (match_def_path(cx.tcx, fn_def_id, &paths::PATH_TO_PATH_BUF)
+                    || match_def_path(cx.tcx, fn_def_id, &paths::OS_STR_TO_OS_STRING));
+
+            if !from_borrow && !from_deref {
+                continue;
+            }
+
+            // _1 in MIR `{ _2 = &_1; clone(move _2); }` or `{ _2 = _1; to_path_buf(_2); } (from_deref)
+            // In case of `from_deref`, `arg` is already a reference since it is `deref`ed in the previous
+            // block.
+            let cloned = unwrap_or_continue!(find_stmt_assigns_to(arg, from_borrow, bbdata.statements.iter().rev()));
+
+            // _1 in MIR `{ _2 = &_1; _3 = deref(move _2); } -> { _4 = _3; to_path_buf(move _4); }`
+            let referent = if from_deref {
+                let ps = mir.predecessors_for(bb);
+                if ps.len() != 1 {
+                    continue;
+                }
+                let pred_terminator = mir[ps[0]].terminator();
+
+                let pred_arg = if_chain! {
+                    if let Some((pred_fn_def_id, pred_arg, pred_arg_ty, Some(res))) =
+                        is_call_with_ref_arg(cx, mir, &pred_terminator.kind);
+                    if *res == mir::Place::Local(cloned);
+                    if match_def_path(cx.tcx, pred_fn_def_id, &paths::DEREF_TRAIT_METHOD);
+                    if match_type(cx, pred_arg_ty, &paths::PATH_BUF)
+                        || match_type(cx, pred_arg_ty, &paths::OS_STRING);
+                    then {
+                        pred_arg
+                    } else {
+                        continue;
+                    }
+                };
+
+                unwrap_or_continue!(find_stmt_assigns_to(pred_arg, true, mir[ps[0]].statements.iter().rev()))
+            } else {
+                cloned
+            };
+
+            let used_later = traversal::ReversePostorder::new(&mir, bb).skip(1).any(|(tbb, tdata)| {
+                // Give up on loops
+                if tdata.terminator().successors().any(|s| *s == bb) {
+                    return true;
+                }
+
+                let mut vis = LocalUseVisitor {
+                    local: referent,
+                    used_other_than_drop: false,
+                };
+                vis.visit_basic_block_data(tbb, tdata);
+                vis.used_other_than_drop
+            });
+
+            if !used_later {
+                let span = terminator.source_info.span;
+                let node = if let mir::ClearCrossCrate::Set(scope_local_data) = &mir.source_scope_local_data {
+                    scope_local_data[terminator.source_info.scope].lint_root
+                } else {
+                    unreachable!()
+                };
+
+                if_chain! {
+                    if !in_macro(span);
```

Comment:
Move this check all the way to the start of the loop and continue if we are in a macro.